src/core/backbone_router/bbr_manager.cpp

```diff
@@ -98,61 +101,104 @@ void Manager::HandleMulticastListenerRegistration(const Coap::Message &aMessage,
-    Ip6::Address               failedAddresses[kIPv6AddressesNumMax];
-    uint8_t                    failedAddressNum = 0;
-    TimeMilli                  expireTime;
     BackboneRouterConfig       config;
 
+    uint16_t     addressesOffset, addressesLength;
+    Ip6::Address address;
+    Ip6::Address failedAddresses[kIPv6AddressesNumMax];
+    uint8_t      failedAddressNum = 0;
+    TimeMilli    expireTime;
+
+    uint32_t timeout;
+    uint16_t commissionerSessionId;
+    bool     hasCommissionerSessionIdTlv = false;
+    bool     hasTimeoutTlv               = false;
+
     VerifyOrExit(aMessage.IsConfirmable() && aMessage.GetCode() == OT_COAP_CODE_POST, error = OT_ERROR_PARSE);
 
     VerifyOrExit(isPrimary, status = ThreadStatusTlv::kMlrBbrNotPrimary);
 
     // TODO: (MLR) send configured MLR response for Reference Device
-    // TODO: (MLR) handle Commissioner Session TLV
-    // TODO: (MLR) handle Timeout TLV
+
+    if (ThreadTlv::FindUint16Tlv(aMessage, ThreadTlv::kCommissionerSessionId, commissionerSessionId) == OT_ERROR_NONE)
+    {
+        const MeshCoP::CommissionerSessionIdTlv *commissionerSessionIdTlv =
+            static_cast<const MeshCoP::CommissionerSessionIdTlv *>(
+                Get<NetworkData::Leader>().GetCommissioningDataSubTlv(MeshCoP::Tlv::kCommissionerSessionId));
+
+        VerifyOrExit(commissionerSessionIdTlv != nullptr &&
+                         commissionerSessionIdTlv->GetCommissionerSessionId() == commissionerSessionId,
+                     error = OT_ERROR_PARSE);
+
+        hasCommissionerSessionIdTlv = true;
+    }
+
+    hasTimeoutTlv = hasCommissionerSessionIdTlv &&
+                    (ThreadTlv::FindUint32Tlv(aMessage, ThreadTlv::kTimeout, timeout) == OT_ERROR_NONE);
 
     VerifyOrExit(ThreadTlv::FindTlvValueOffset(aMessage, IPv6AddressesTlv::kIPv6Addresses, addressesOffset,
                                                addressesLength) == OT_ERROR_NONE,
                  error = OT_ERROR_PARSE);
     VerifyOrExit(addressesLength % sizeof(Ip6::Address) == 0, error = OT_ERROR_PARSE);
     VerifyOrExit(addressesLength / sizeof(Ip6::Address) <= kIPv6AddressesNumMax, error = OT_ERROR_PARSE);
 
-    SuccessOrExit(error = Get<BackboneRouter::Leader>().GetConfig(config));
-    expireTime = TimerMilli::GetNow() + TimeMilli::SecToMsec(config.mMlrTimeout);
+    if (!hasTimeoutTlv)
+    {
+        SuccessOrExit(error = Get<BackboneRouter::Leader>().GetConfig(config));
 
-    for (uint16_t offset = 0; offset < addressesLength; offset += sizeof(Ip6::Address))
+        timeout = config.mMlrTimeout;
+    }
+    else
     {
-        bool failed = true;
+        VerifyOrExit(timeout < std::numeric_limits<uint32_t>::max(), status = ThreadStatusTlv::kMlrNoPersistent);
+
+        if (timeout != 0)
+        {
+            timeout = std::min(timeout, static_cast<uint32_t>(Mle::kMlrTimeoutMax));
+            timeout = std::max(timeout, static_cast<uint32_t>(Mle::kMlrTimeoutMin));
```

Comment:
In theory, the `timeout` value in `TimeoutTlv` should not be changed in the recipient I think.
When changing it, better to add some log?

Besides, is it necessary to require it `>= kMlrTimeoutMin` here?
For MLR from general devices, it was already pre-processed when getting PBBR Dataset.
For MLR from commissioner which is allowed to take TimeoutTlv, could we simply respect the value inside?

Reply:
Here the code handles TimeoutTLV which is for Commissioners only.
Even if we allow Commissioner to configure Timeout TLV < `kMlrTimeoutMin`, it's useless because devices will treat `MLR Timeout < kMlrTimeoutMin` as `kMlrTimeoutMin` which is what Thread Spec. requires. 
So, I would prefer to avoid such "invalid" configuration to be applied in the first place.

It's a good idea to add a log here.